core/sync/theil_sen.go

```diff
@@ -0,0 +1,127 @@
+package sync
+
+import (
+	"sort"
+	"time"
+
+	"go.uber.org/zap"
+
+	"example.com/scion-time/base/timebase"
+)
+
+type theilSen struct {
+	log      *zap.Logger
+	clk      timebase.LocalClock
+	pts      []timePoint
+	baseFreq float64
+}
+
+// If the buffer size is too large, the system is likely to oscillate heavily.
+const MeasurementBufferSize = 4
+
+const BaseFreqGainFactor = 1 / (0.005)
+
+func newTheilSen(log *zap.Logger, clk timebase.LocalClock) *theilSen {
+	return &theilSen{log: log, clk: clk, pts: make([]timePoint, 0), baseFreq: 0.0}
+}
+
+type timePoint struct {
+	x time.Time
+	y time.Time
+}
+
+type point struct {
+	x int64
+	y int64
+}
+
+func median(ds []float64) float64 {
+	n := len(ds)
+	if n == 0 {
+		panic("unexpected number of slopes")
+	}
+
+	sort.Float64s(ds)
+
+	var m float64
+	i := n / 2
+	if n%2 != 0 {
+		m = ds[i]
+	} else {
+		m = ds[i-1] + (ds[i]-ds[i-1])/2
+	}
+	return m
+}
+
+func slope(inputs []point) float64 {
+	if len(inputs) == 1 {
+		return float64(inputs[0].y) / float64(inputs[0].x)
+	}
+
+	var medians []float64
+	for idxA, pointA := range inputs {
+		for _, pointB := range (inputs)[idxA+1:] {
+			// Like in the original paper by Sen (1968), ignore pairs with the same x coordinate
+			if pointA.x != pointB.x {
+				medians = append(medians, float64(pointA.y-pointB.y)/float64(pointA.x-pointB.x))
+			}
+		}
+	}
+
+	if len(medians) == 0 {
+		panic("invalid inputs: all inputs have the same x coordinate")
```

Comment:
`len(pts) == 0` also leads to this panic, right? Maybe the function theilSen.Offset should return an additional value `ok bool` indicating whether an offset is currently available.

Reply:
As for the case `same x coordinate`: Wouldn't it be possible to detect this case already in `AddSample`? To keep the structure simple, we could panic there with "unexpected clock behavior" like in pll.go.

Reply:
I think it would be better to try and detect this case in e.g. `regressionPts` rather than `AddSample`. I'm considering the following scenario:

If we eventually do not always immediately call `Offset` directly after `AddSample` anymore, which might make sense, it could be that we first add two samples with the same x coordinate. Then `AddSample` would panic, since all samples have the same x. But if then another sample with a different x coordinate was added before `Offset` was called, we would not have had to panic.